compiler/src/main/kotlin/walkers/type.kt

```diff
@@ -0,0 +1,333 @@
+package dk.aau.cs.d409f19.cellumata.walkers
+
+import dk.aau.cs.d409f19.cellumata.ast.*
+import kotlin.AssertionError
+
+/**
+ * Thrown when a violation of the type rules is found
+ */
+class TypeError : Exception()
+
+/**
+ * Move types up the abstract syntax tree according to the type rules, and check that there is no violation of the type rules
+ */
+class TypeChecker(symbolTable: Table) : ScopedASTVisitor(symbolTable = symbolTable) {
+    override fun visit(node: OrExpr) {
+        super.visit(node)
+
+        if (node.left.getType() != BooleanType) {
+            throw TypeError()
+        }
+        if (node.right.getType() != BooleanType) {
+            throw TypeError()
+        }
+
+        node.setType(BooleanType)
+    }
+
+    override fun visit(node: AndExpr) {
+        super.visit(node)
+
+        if (node.left.getType() != BooleanType) {
+            throw TypeError()
+        }
+        if (node.right.getType() != BooleanType) {
+            throw TypeError()
+        }
+
+        node.setType(BooleanType)
+    }
+
+    override fun visit(node: InequalityExpr) {
+        super.visit(node)
+
+        node.setType(when {
+            node.left.getType() == BooleanType && node.right.getType() == BooleanType -> BooleanType
+            node.left.getType() == IntegerType && node.right.getType() == IntegerType -> BooleanType
+            node.left.getType() == FloatType && node.right.getType() == FloatType -> BooleanType
+            node.left.getType() == IntegerType && node.right.getType() == FloatType -> BooleanType
+            node.left.getType() == FloatType && node.right.getType() == IntegerType -> BooleanType
+            node.left.getType() == StateType && node.right.getType() == StateType -> BooleanType
+            node.left.getType() == ActualNeighbourhoodType && node.right.getType() == ActualNeighbourhoodType -> BooleanType
+            node.left.getType() == ActualNeighbourhoodType && node.right.getType() is ArrayType && (node.right.getType() as ArrayType).subtype == StateType -> BooleanType
+            node.left.getType() is ArrayType && (node.left.getType() as ArrayType).subtype == StateType && node.right.getType() == ActualNeighbourhoodType -> BooleanType
```

Comment:
Hella long lines here and below, linebreak possible?

Reply:
Suggestions are welcome

Reply:
We may want to think about a distributed codestyle for IntelliJ, but not now. It's fine as is.